### ISSUE TITLE
We should drop animation and Voronoi, and focus on making react-d3 actually be d3 on React

### ISSUE THREAD

**Issue by MatthewHerbst** [esbullington/react-d3#223]
Implementing our own animations is:

- preventing us from making this library what is should be - d3.js implemented in a React.js context
  - d3.js doesn't do animation for you - it just empowers you to add it in 
    - small caveat with `d3.transition()`, but what we are doing is a far cry from that
  - spending time on animation is time not spent adding in more d3.js features
- creating bugs and problems that don't need to be created
- making the code base larger and significantly harder to maintain (and to understand for new contributors)
- preventing chart users from using all the features charts should have in a uniform manner
  - the implementation of voronoi is flawed in that it cherry picks data provided by the user
  - voronoi also has many bugs just waiting to happen (for example, it doesn't like "NaN", so instead you implicitly return "undefined" which will also cause the code to fail)
  - users have no power to interact with their charts
- causing the library to deviate from actually using d3.js features. Just look at how the path is drawn in `ScatterChart`

I think that we should strip all the animation and voronoi code from the library (sorry @yang-wei , I know you've spent a crap ton of time on it - I don't mean to insult you in any way; it's a good idea, but I truly think it holds the library back).

If you want users to be able to do quick animations, then setup a demo gallery of charts that implement animations *the right way (event callbacks)*. This is similar to what @mbostock has done with d3.js; he has created a very powerful library and then has a bunch of examples of how to use that library, but he doesn't include any of that stuff in the actual code base itself.

I realize this is probably dead-on-arrival to you guys, but again, I really think it is the right path. This library isn't "easy-react-charts". It's "react-d3". So, we should actually make it be d3.js - on React.

(All this comes from working through some stuff with `ScatterChart` and `utils.flattenData` and realizing that voronoi and `utils.flattenData`, in their current forms, are, in my humble opinion, not worth saving. For example, `ScatterChart` is not using `props.colorAccessor`, but instead has implemented its own in the form of `(d, idx) => d)`. Try changing it to `props.colorAccessor` and the tests fail. Realize why the tests are failing, look at the changes that will be needed, and I think you will agree with me.)

**Comment by MatthewHerbst** [comment]
@esbullington haven't heard from you on anything for a while now. I know you're busy, but your thoughts on all of this would be greatly appreciated.

**Comment by parshap** [comment]
> I realize this is probably dead-on-arrival to you guys, but again, I really think it is the right path. This library isn't "easy-react-charts". It's "react-d3". So, we should actually make it be d3.js - on React.

This is not dead-on-arrival to *me*. I agree, we should try to expose d3's semantics and not much more.

I am, however not opposed to having something like `react-d3/easy-charts` where we bundle things like high-level charts and animation helpers. This would give us real-world use cases bundled in the same codebase, helping make sure the low-level react-d3 module continues to serve its purpose.

**Comment by MatthewHerbst** [comment]
@parshap I couldn't have said it better myself :+1:

**Comment by esbullington** [comment]
tldr: d3 is a dataviz library, this is a chart library.  Look to MetricsGraphics for inspiration (also build on d3)

To answer the question specifically, this is meant to be "d3-easy-charts".  On retrospect, I should have chosen another name, but I'm far from the only one who has made the mistake of naming a charting library based on "d3" (see nvd3, c3, etc.).  And by the time I realized it, there were already a number of companies using the project in production.

In fact, I had planned to also create a lower-level library that mirrored some of d3's functionality in React, but the fact is, at this time it's not even possible to do so.  Until React exposes some lower-level animation API, we *can't* mirror d3's functionality at even a basic level.  I know that chenglou and others at Facebook are working on this, and when they have something ready, I may revisit the issue and look into building a lower-level library more like d3, and then bundle off the individual charts into their own repos.

That said, I *am* currently interested in a lower-level API for building *charts*, which you can read about [here](https://github.com/esbullington/react-d3/pull/9#issuecomment-66976230).  That's basically how we implement our charts internally, but I wanted to polish it up and document it for external use.  *This* is the API we should use direct advanced users to, instead of piling on additional properties on the top-level charts.  For basic users, and people who just want a chart that works, the `<FooChart/>` API should be as simple, well-documented, and bug-free as possible.

At this time, my priorities are:

1.  Eliminating the bugs.
2.  Documenting features (wiki is open for anyone to contribute)
3.  Supporting users currently using `react-d3` in production, and listening to them for guidance on features needed (e.g., @samdalton and @parshap among others).  A big need is self-scaling charts, which we're real close to having (samdalton and parshap have good insight into this, and have implemented it in their fork).
4.  Fixing and simplifying animation  @yang-wei has a good handle on this
5.  Putting out a minor release when the project is as bug-free as possible (@yang-wei will let me know when that is)

I'm sorry I've not been able to spend more time with the project, but my new job will be extremely busy until the end of June (I *like* my new job, so I want to do well).  In the meantime, please look to @yang-wei for guidance.  He's been a co-contributor since early on, and is responsible for the initial implementation of almost half the charts.  He's got a very good idea of what I had in mind when I create `react-d3`, and some ideas of his own.

**Comment by MatthewHerbst** [comment]
Thanks for the feedback @esbullington The interesting thing about the discussion you have in #9 is that it would allow each `DataSeries` to potentially have unique styling/animations. It seems though that that approach was never implemented? Is that something you would like someone to work on?

As for your priorities, I think the fastest way to eliminate bugs would be to get the animation code working correctly in one of the Charts before it is applied everywhere else. Most of the bugs I have come across are directly related to animation code.

Also: I'm not sure if there is anything to be done about it, but it seems to me that there is a massive amount of code duplication between charts. Is that something that you think can be generalized, or say, does changing some animation code mean we have to go touch every chart (which increases the changes of introducing new bugs)?

**Comment by yang-wei** [comment]
Since @esbullington is busy let me tell you what I know. No Eric is not working on that issue and that's not our main focus. This feature is a great-to-have feature in the future but it will bring up a lot of API changes.

Thanks to your pr we now have props like `colorAccessor` which increase the flexibility to make styling. So I think that's not our main concern now.

**Comment by parshap** [comment]
@esbullington: Thanks for sharing your thoughts and links to the other relevant posts. You have obviously given this topic a lot of thought and I am totally onboard with this library being *d3-easy-charts*.

> In fact, I had planned to also create a lower-level library that mirrored some of d3's functionality in React, but the fact is, at this time it's not even possible to do so. Until React exposes some lower-level animation API, we can't mirror d3's functionality at even a basic level.

Is it necessary to mirror *all* of d3's functionality in React? I think we just need enough to be able to accomplish server-side rendering and to expose an idiomatic React interface (i.e., components and props). I think it would be ok to let d3 to manipulate the dom for things like animations by having `componentDidMount` hooks. Admittedly, I haven't tried to implement this, so I'm not sure if the approach would work in practice.

> That said, I am currently interested in a lower-level API for building charts

A lower-level API for building charts would be great! For example, we've been wanting to create a chart with both a LineChart `<DataSeries>` and an AreaChart `<DataSeries>`. Or two AreaChart `<DataSeries>` layered on top of each other. I also wish we could have callbacks for interactions with charts (e.g., when a data point is clicked). An approach I've been thinking about is rendering our own voronois with custom interaction handlers on top of the DataSeries. Exposing primitives to help do these kinds of things would be awesome.

> I'm sorry I've not been able to spend more time with the project, but my new job will be extremely busy until the end of June (I like my new job, so I want to do well).

Happy to hear about your job! I don't think you need to apologize. You have contributed a great deal already and have no further obligation.
